### DISCUSSION TITLE
workflow is not defined in the config file

### DISCUSSION THREAD

**Discussion by zhilizheng** [nextflow-io/nextflow#4152] [Q&A]
## Bug report 

I tried to get the `${workflow.workDir}` in the config file for trace path and name. Note the `workDir` and `runName` are specified by the nextflow run command line  `-w test -name test`

```nextflow
trace{
    enabled = true
    file = "${workflow.workDir}/reports/${workflow.runName}.trace.txt"
}
```

### Expected behavior and actual behavior

Expected:  get the current work dir and `runName` specified by the command line
Actual:  get an error:
```
ERROR ~ Unknown config attribute `workflow.workDir` -- check config file
```

### Steps to reproduce the problem

Any job with the trace in config file, and specify the `workDir` and name by command line. 

### Program output 

```
ERROR ~ Unknown config attribute `workflow.workDir` -- check config file:
```

### Environment 

* Nextflow version: [23.04.2] 
* Java version: [17.0.8]
* Operating system: [Debian12]
* Bash version: (5.2.15)

**Comment by mribeirodantas** [comment]
One way to achieve such a thing in config files is through closures. You can get `workflow.runName` within a closure for `publishDir.path`. For example, check [here](https://github.com/CenterForMedicalGeneticsGhent/nf-cmgg-germline/blob/1e8a16e710d405ba1f64bfe7d28b36c5c97fcb71/conf/modules.config#L13-L18). Unfortunately, `trace.file` doesn't evaluate closures 😢

@bentsherman

**Comment by pditommaso** [comment]
You can use a parameter to specify the work dir and reference in other settings 

```
params.workDir = '/some/path'

workDir = params.workDir

trace{
    enabled = true
    file = "${params.workDir}/reports/etc.trace.txt"
}
```

**Reply by mateuszpawlik** [reply]
And what if I'd like to overwrite it from the command line? I'm trying but I can't make it work.

**Comment by stevekm** [comment]
hey just wanted to mention, I had similar issues as well recently, I threw together this little demo that shows different ways to dig out Nextflow implicit variables in a few different contexts, including within the nextflow.config, might be helpful for this discussion https://github.com/stevekm/nextflow-demos/tree/master/workflow-introspection
